File: doughmination.win/backend/app/api/system.py
"""
System information API endpoints
Handles system info and mental state
"""

from fastapi import APIRouter, HTTPException, Depends

from app.models import MentalState
from app.services.pluralkit import get_system
from app.services.mental_state_service import get_mental_state, save_mental_state
from app.dependencies.auth import get_admin_user
from app.core.websocket import broadcast_mental_state_update

router = APIRouter()

@router.get("/system")
async def system_info():
    """Get PluralKit system information with mental state"""
    try:
        # Get system data from PluralKit
        system_data = await get_system()
        
        # Get mental state
        mental_state = get_mental_state()
        
        # Add mental state to system data
        system_data["mental_state"] = mental_state.dict()
        
        return system_data
        
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch system info: {str(e)}"
        )

@router.get("/mental-state")
async def get_mental_state_endpoint():
    """Get current mental state"""
    try:
        state = get_mental_state()
        return state
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to fetch mental state: {str(e)}"
        )

@router.post("/mental-state")
async def update_mental_state_endpoint(
    state: MentalState,
    user=Depends(get_admin_user)
):
    """Update mental state (admin only)"""
    try:
        success = save_mental_state(state)
        
        if not success:
            raise HTTPException(
                status_code=500,
                detail="Failed to save mental state"
            )
        
        # Broadcast the mental state update
        state_data = state.dict()
        state_data["updated_at"] = state_data["updated_at"].isoformat()
        await broadcast_mental_state_update(state_data)
        
        return {"success": True, "message": "Mental state updated"}
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update mental state: {str(e)}"
        )

File: system/backend/src/routes/system.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * System information API endpoints.
 * Handles system info and mental state.
 *
 * Note: FastAPI validates the `state: MentalState` request body
 * automatically before the handler runs, returning 422 on failure.
 * Express has no equivalent, so the POST handler below validates with
 * MentalStateSchema.safeParse() up front and returns 422 itself to match.
 */

import { Router, type Request, type Response, type NextFunction } from 'express';

import { MentalStateSchema } from '../models/index.js';
import { getSystem } from '../services/pluralkit.js';
import { getMentalState, saveMentalState } from '../services/mental_state_service.js';
import { requireAuth, requireAdmin } from '../dependencies/auth.js';
import { broadcastMentalStateUpdate } from '../core/websocket.js';
import { HttpError } from '../core/errors.js';

export const systemRouter = Router();

/** Get PluralKit system information with mental state */
systemRouter.get('/system', async (_req: Request, res: Response, next: NextFunction) => {
  try {
    // Get system data from PluralKit
    const systemData = await getSystem();

    // Get mental state
    const mentalState = await getMentalState();

    // Add mental state to system data
    systemData.mental_state = mentalState;

    res.json(systemData);
  } catch (err) {
    next(new HttpError(500, `Failed to fetch system info: ${String(err)}`));
  }
});

/** Get current mental state */
systemRouter.get('/mental-state', async (_req: Request, res: Response, next: NextFunction) => {
  try {
    const state = await getMentalState();
    res.json(state);
  } catch (err) {
    next(new HttpError(500, `Failed to fetch mental state: ${String(err)}`));
  }
});

/** Update mental state (admin only) */
systemRouter.post(
  '/mental-state',
  requireAuth,
  requireAdmin,
  async (req: Request, res: Response, next: NextFunction) => {
    const parsed = MentalStateSchema.safeParse(req.body);
    if (!parsed.success) {
      res.status(422).json({ detail: parsed.error.issues });
      return;
    }
    const state = parsed.data;

    try {
      const success = await saveMentalState(state);

      if (!success) {
        next(new HttpError(500, 'Failed to save mental state'));
        return;
      }

      // Broadcast the mental state update
      const stateData = { ...state, updated_at: state.updated_at.toISOString() };
      await broadcastMentalStateUpdate(stateData);

      res.json({ success: true, message: 'Mental state updated' });
    } catch (err) {
      next(new HttpError(500, `Failed to update mental state: ${String(err)}`));
    }
  },
);